### ISSUE TITLE
Misleading first level parameters logic

### ISSUE THREAD

**Issue by vferat** [nilearn/nilearn#5408]
The current handling of the `slice_time_ref` parameter in `nilearn.glm.first_level.FirstLevelModel` can be misleading.

Although `slice_time_ref` is defined at the class initialization level, it is only used within the `fit` method **if** `design_matrices` is **not** provided. This creates a confusing situation where:

* Users may expect `slice_time_ref` to always be taken into account once set during instantiation.
* In practice, the parameter has **no effect** when precomputed design matrices are passed to `.fit()`.

I think it would be clearer if:

- the `slice_time_ref` parameter would move to the `fit` method signature (since it's only relevant there).
- Use `slice_time_ref` even when `design_matrices` are provided (e.g., by applying the same adjustment logic).

or 
-  `slice_time_ref` is an optional argument in `make_first_level_design_matrix`.
 - a **warning** in the documentation and/or during the execution of `.fit()` are displayed to inform users that `slice_time_ref` is ignored when `design_matrices` is manually supplied.

I understand that changing the function signature or behavior may introduce backward compatibility issues. However, given the potential for silent misuse, I believe it's worth addressing (even if through warnings ) to help users avoid incorrect assumptions in their analyses.

This issue is not linked to a specific platform nor a specific python/nilearn version.

**Comment by bthirion** [comment]
Thx for opening this. To clarify:
* Following scikit-learn's logic, we do not provide model specification arguments in fit, as fit is simply supposed to take some data and run the analysis on these data, with parameters fixed to their instantiation values. I believe this behavior is preferable for the sake of pipeline validity, and it is a standard we are reluctant to give up. 
* Applying a slice time ref difference when design matrices are provided is problematic for two reasons: 1) we would need to know the time reference used to generate the matrices ;  we would also need to know which columns the resampling should be applied to; 2) temporal resampling requires additional parameter specification, such as the resampling method (sinc ? linear ?). So I'd rather have a specific function that does it outside of the GLM model.

However, I understand the issue you encounter. Maybe adding a warning, as you suggest, is a good thing to do. What do others think ?

**Comment by Remi-Gau** [comment]
from the sklearn doc: https://scikit-learn.org/stable/developers/develop.html#fitting

> Depending on the nature of the algorithm, fit can sometimes also accept additional keywords arguments. However, any parameter that can have a value assigned prior to having access to the data should be an __init__ keyword argument. Ideally, fit parameters should be restricted to directly data dependent variables.

so in that sense one 'could' make the argument that some parameters (slice_time_ref and t_r) passed at init could be passed at fit as they are data dependent

in any case I agree that we should make it 'in your face' obvious when some parameters are not used because the code rely on passed desgin matrices

however to be consistent, we should do this for ALL the parameters that are then ignored:
- slice_time_ref
- t_r
- hrf_model
- drift_model
- drift_order
- fir_delays
- min_onset

https://github.com/nilearn/nilearn/blob/3b38e6372129950a9b2146bac9a296a8ea6e326f/nilearn/glm/first_level/first_level.py#L687-L701

```
        start_time = self.slice_time_ref * self.t_r
        end_time = (n_scans - 1 + self.slice_time_ref) * self.t_r
        frame_times = np.linspace(start_time, end_time, n_scans)
        design = make_first_level_design_matrix(
            frame_times,
            events[run_idx],
            self.hrf_model,
            self.drift_model,
            self.high_pass,
            self.drift_order,
            self.fir_delays,
            confounds_matrix,
            confounds_names,
            self.min_onset,
        )
```

for now I would say let's update the doc and add warnings and keep discussing to see how we feel about changing the API

**Comment by Remi-Gau** [comment]
OK another advantage to pass slice_time_ref and t_r ad fit time is that this can allow different values for each fitted run: definitely a niche use case where user wants to run the same model on runs that have been acquired / preprocessed differently (of course this is sill doable in any wase)

**Comment by emdupre** [comment]
I'm against changing the API, but I think a warning when providing both `design_matrices` and ignored params is reasonable. Right now the documentation notes that `design_matrices` takes precedence over events and confounds, but this is clearly not explicit enough.

**Comment by bthirion** [comment]
> OK another advantage to pass slice_time_ref and t_r ad fit time is that this can allow different values for each fitted run: definitely a niche use case where user wants to run the same model on runs that have been acquired / preprocessed differently (of course this is sill doable in any wase)

But people will mess up with that. I find it rather error-prone.

**Comment by Remi-Gau** [comment]
> But people will mess up with that. I find it rather error-prone.

yeah I was more trying to play devil's advocate and see what possible pros there could be.

overall I am more on the let's not change the API side

**Comment by Remi-Gau** [comment]
was decided at the last dev meeting to not change the API and to stick to add warnings and improve the documentation
